src/Build/BuildCheck/Infrastructure/CustomConfigurationData.cs

```diff
@@ -55,10 +65,40 @@ public override bool Equals(object? obj)
             return false;
         }
 
-        return Equals((CustomConfigurationData)obj);
-    }
+        var customConfigObj = (CustomConfigurationData) obj;
+
+        if(customConfigObj.RuleId != RuleId)
+        {
+            return false;
+        }
 
-    protected bool Equals(CustomConfigurationData other) => Equals(ConfigurationData, other.ConfigurationData);
+        // validate keys and values
+        if (customConfigObj.ConfigurationData != null && ConfigurationData != null)
+        {
+            if (!customConfigObj.ConfigurationData.Keys.SequenceEqual(ConfigurationData.Keys))
+            {
+                return false;
+            }
+
+            var keys = customConfigObj.ConfigurationData.Keys;
+            foreach (var key in keys)
+            {
+                if (customConfigObj.ConfigurationData[key] != ConfigurationData[key])
+                {
+                    return false;
+                }
+            }
```

Comment:
nit: It's faster to enumerate key-value pairs.

```suggestion
            foreach (var kvp in customConfigObj.ConfigurationData)
            {
                if (kvp.Value != ConfigurationData[kvp.Key])
                {
                    return false;
                }
            }
```

Reply:
this is updated, thank you!